src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -750,6 +750,58 @@ void resetTypesShapeToQuestionmarks(Operation *op) {
     resetTypeShapeToQuestionmarks(op->getResult(i));
 }
 
+//===----------------------------------------------------------------------===//
+// ONNX Custom Op Shape Helper
+//===----------------------------------------------------------------------===//
+
+ONNXCustomOpShapeHelper::ONNXCustomOpShapeHelper(Operation *op,
+    ValueRange operands, IndexExprBuilder *ieBuilder, IndexExprScope *scope,
+    bool hasUniBroadcasting)
+    : ONNXOpShapeHelper(op, operands, ieBuilder, scope),
+      ONNXUnaryOpShapeHelper(op, operands, ieBuilder, scope),
+      ONNXBroadcastOpShapeHelper(op, operands, ieBuilder, scope),
+      operandsVector() {
+  ONNXCustomOp customOp = cast<ONNXCustomOp>(op);
+  if (!customOp.getShapeInferPattern().has_value()) {
+    pattern = 0;
+    return;
+  } else if (customOp.getShapeInferPattern() == "SameAs") {
+    pattern = 1;
+  } else if (customOp.getShapeInferPattern() == "MDBroadcast") {
+    pattern = 2;
+  } else {
+    // Or just use ShapeHelper?
+    llvm_unreachable("The specified shape_infer_pattern is not supported"
+                     "Error encountered in shape inference.");
+  }
+  std::optional<ArrayAttr> inputIndexAttrs = customOp.getInputsForInfer();
+
+  ValueRange inputs =
+      operands.empty() ? ValueRange(customOp.getInputs()) : operands;
+  if (!inputIndexAttrs.has_value()) {
+    return;
+  }
+
+  // ToFix: It seems to me that the vector for ValueRange has to be alive
+  // during the lifespan of the ValueRange. Using static definitely violates
+  // MLIR transformation requirement.
+
+  ValueRange usedOperands;
+  for (auto indexAttr : inputIndexAttrs.value()) {
+    operandsVector.push_back(inputs[indexAttr.cast<IntegerAttr>().getInt()]);
+  }
+  ONNXOpShapeHelper::operands = ArrayRef(operandsVector);
+}
+
+LogicalResult ONNXCustomOpShapeHelper::computeShape() {
+  if (pattern == 1) {
+    return ONNXUnaryOpShapeHelper::computeShape();
```

Comment:
When the "SameAs" pattern was defined, I thought that there was an option to define the index of the input it was the same as. Is this supported here?

Reply:
In fact, there is no need to do double inheritance. You can have the custom op as a child of Broadcast shape inference, as the call that you really need is this one
```
LogicalResult ONNXOpShapeHelper::setOutputDimsFromOperand(
    Value operand, int n, bool refineShape) {
```
which is defined on the parent class already. So you can simply pick up one of the "id" that reflect the same as shape, andset the output shape using this. And when its is a broadcast, then use the super (which is broadcast shape inf).

Reply:
The reason for double inheritance is that the computeShape() is a pure virtual function. I think that our code pattern is always call computeShape() first. And the computeShape for UnaryOp is just setOutputDimsFromOperand(operands[0]).
My early version of this PR did not use computeShape() and exposed the logic of computeShape in the code. I was asked by @AlexandreEichenberger to follow the existing code pattern. Now should I just change back to my previous implementation?
Another point is that the SameAs is not the same as UnaryOpShapeHelper.  The SameAs with more than one inputs could choose to propagate shape info from the best one and allow dim symbolic analysis to unify symbols. 
@AlexandreEichenberger @tungld Should we schedule a meeting to discuss this topic?